Client requests not using RequestHeaders.authority

Not sure if this is intentional, but it was definitely unexpected

Minimal reproducer

```
final RequestHeadersBuilder requestHeadersBuilder =
                RequestHeaders.builder()
                              .authority("localhost:8080")
                              .scheme("h2c")
                              .add("param1", "val1")
                              .path("/path1");

        final AggregatedHttpRequest request = AggregatedHttpRequest.of(
                requestHeadersBuilder.method(HttpMethod.GET).build());
        final AggregatedHttpResponse response = WebClient.of().execute(request).aggregate().join();
```

Logs
```
Caused by: java.lang.IllegalArgumentException: no authority: /path1
	at com.linecorp.armeria.client.DefaultWebClient.execute(DefaultWebClient.java:70)
	at com.linecorp.armeria.client.DefaultWebClient.execute(DefaultWebClient.java:101)
	... 64 more
```

--- a/core/src/main/java/com/linecorp/armeria/client/DefaultWebClient.java
+++ b/core/src/main/java/com/linecorp/armeria/client/DefaultWebClient.java
@@ -54,7 +54,11 @@ final class DefaultWebClient extends UserClient<HttpRequest, HttpResponse> imple
                 uri = null;
             }
         } else {
-            uri = null;
+            if (req.scheme() != null && req.authority() != null) {
+                uri = req.uri();
+            } else {
+                uri = null;
+            }
         }
 
         if (uri != null) {

--- a/core/src/test/java/com/linecorp/armeria/client/WebClientBuilderTest.java
+++ b/core/src/test/java/com/linecorp/armeria/client/WebClientBuilderTest.java
@@ -19,8 +19,31 @@ package com.linecorp.armeria.client;
 import static org.assertj.core.api.Assertions.assertThat;
 
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import com.linecorp.armeria.common.AggregatedHttpRequest;
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.RequestHeaders;
+import com.linecorp.armeria.common.RequestHeadersBuilder;
+import com.linecorp.armeria.server.ServerBuilder;
+import com.linecorp.armeria.testing.junit.server.ServerExtension;
 
 class WebClientBuilderTest {
+
+    @RegisterExtension
+    static ServerExtension server = new ServerExtension() {
+        @Override
+        protected void configure(ServerBuilder sb) throws Exception {
+            sb.service("/echo-path", (ctx, req) -> {
+                String pathAndQuery = ctx.path();
+                if (ctx.query() != null) {
+                    pathAndQuery += '?' + ctx.query();
+                }
+                return HttpResponse.of(pathAndQuery);
+            });
+        }
+    };
 
     @Test
     void uriWithNonePlusProtocol() throws Exception {
@@ -46,6 +69,23 @@ class WebClientBuilderTest {
         final WebClient client = WebClient.builder("http", Endpoint.of("127.0.0.1"), "/foo")
                                           .build();
         assertThat(client.uri().toString()).isEqualTo("http://127.0.0.1/foo");
+    }
+
+    @Test
+    void authorityHeader() {
+        final String path = "/echo-path?foo=bar";
+        final RequestHeadersBuilder requestHeadersBuilder =
+                RequestHeaders.builder()
+                              .authority("localhost:" + server.httpPort())
+                              .scheme("h2c")
+                              .add("param1", "val1")
+                              .path(path);
+
+        final AggregatedHttpRequest request = AggregatedHttpRequest.of(
+                requestHeadersBuilder.method(HttpMethod.GET).build());
+        final HttpResponse response = WebClient.of().execute(request);
+        assertThat(response.isOpen()).isTrue();
+        assertThat(response.aggregate().join().contentUtf8()).isEqualTo(path);
     }
 
     @Test

